pkg/proxy/handlers/spacelister_get_test.go

```diff
@@ -726,3 +742,234 @@ func TestGetUserWorkspace(t *testing.T) {
+			s := &handlers.SpaceLister{
+				GetSignupFunc:          signupProvider,
+				GetInformerServiceFunc: informerFunc,
+				ProxyMetrics:           proxyMetrics,
+			}
+
+			e := echo.New()
+			req := httptest.NewRequest(http.MethodGet, "/", strings.NewReader(""))
+			rec := httptest.NewRecorder()
+			ctx := e.NewContext(req, rec)
+			ctx.Set(rcontext.UsernameKey, tc.username)
+			ctx.Set(rcontext.RequestReceivedTime, time.Now())
+			ctx.SetParamNames("workspace")
+			ctx.SetParamValues(tc.workspaceRequest)
+			ctx.Set(rcontext.PublicViewerEnabled, true)
+
+			// when
+			wrk, err := handlers.GetUserWorkspace(ctx, s, tc.workspaceRequest)
+
+			// then
+			if tc.expectedErr != "" {
+				// error case
+				require.Error(t, err, tc.expectedErr)
+			} else {
+				require.NoError(t, err)
+			}
+
+			if tc.expectedWorkspace != nil {
+				require.Equal(t, tc.expectedWorkspace, wrk)
+			} else {
+				require.Nil(t, wrk) // user is not authorized to get this workspace
+			}
+		})
+	}
+}
+
+func TestSpaceListerGetWithBindingsWithPublicViewerEnabled(t *testing.T) {
+
+	fakeSignupService := fake.NewSignupService(
+		newSignup("batman", "batman.space", true),
+		newSignup("robin", "robin.space", true),
+	)
+
+	fakeClient := fake.InitClient(t,
+		// NSTemplateTiers
+		fake.NewBase1NSTemplateTier(),
+
+		// space
+		fake.NewSpace("robin", "member-1", "robin"),
+		fake.NewSpace("batman", "member-1", "batman"),
+
+		// spacebindings
+		fake.NewSpaceBinding("robin-1", "robin", "robin", "admin"),
+		fake.NewSpaceBinding("batman-1", "batman", "batman", "admin"),
+
+		fake.NewSpaceBinding("community-robin", toolchainv1alpha1.KubesawAuthenticatedUsername, "robin", "viewer"),
+	)
+
+	robinWS := workspaceFor(t, fakeClient, "robin", "admin", true,
+		commonproxy.WithBindings([]toolchainv1alpha1.Binding{
+			{
+				MasterUserRecord: toolchainv1alpha1.KubesawAuthenticatedUsername,
+				Role:             "viewer",
+				AvailableActions: []string{},
+			},
+			{
+				MasterUserRecord: "robin",
+				Role:             "admin",
+				AvailableActions: []string{},
+			},
+		}),
+		commonproxy.WithAvailableRoles([]string{"admin", "viewer"}),
+	)
+
+	batmanWS := workspaceFor(t, fakeClient, "batman", "admin", true,
+		commonproxy.WithBindings([]toolchainv1alpha1.Binding{
+			{
+				MasterUserRecord: "batman",
+				Role:             "admin",
+				AvailableActions: []string{},
+			},
+		}),
+		commonproxy.WithAvailableRoles([]string{"admin", "viewer"}),
+	)
+
+	tests := map[string]struct {
+		username          string
+		workspaceRequest  string
+		expectedWorkspace *toolchainv1alpha1.Workspace
+	}{
+		"robin can get robin workspace": {
+			username:          "robin.space",
+			workspaceRequest:  "robin",
+			expectedWorkspace: &robinWS,
+		},
+		"batman can get batman workspace": {
+			username:          "batman.space",
+			workspaceRequest:  "batman",
+			expectedWorkspace: &batmanWS,
+		},
+		"batman can get robin workspace": {
```

Comment:
Same here as above. Let's add a test for the case when the public viewer feature is disabled in the context so batman can't get the robin workspace. Unless it's already covered somehow and I missed it.

Reply:
done in a525db583e260bde5f677a42e6c1ebe3906c1e12